File: native/ios/VocaKit/Sources/VocaKit/Screens/SettingsScreen.swift
import SwiftUI

/// JSX `SettingsScreen`: session toggles, subscription, offline languages with
/// install buttons, legal links.
public struct SettingsScreen: View {
    @ObservedObject var preferences: Preferences
    @ObservedObject var entitlements: EntitlementStore
    let offline: OfflineAvailability
    let privacyURL: URL
    let onUpgrade: () -> Void
    let onStartOffline: () -> Void

    public init(preferences: Preferences, entitlements: EntitlementStore, offline: OfflineAvailability, privacyURL: URL,
                onUpgrade: @escaping () -> Void, onStartOffline: @escaping () -> Void) {
        self.preferences = preferences
        self.entitlements = entitlements
        self.offline = offline
        self.privacyURL = privacyURL
        self.onUpgrade = onUpgrade
        self.onStartOffline = onStartOffline
    }

    public var body: some View {
        ScrollView {
            VStack(spacing: 0) {
                ScreenTitle("Settings")
                VStack(alignment: .leading, spacing: 20) {
                    group("Session") {
                        ToggleRow("Automated notes", hint: "Summarise each session when it ends", isOn: $preferences.notes)
                        divider
                        ToggleRow("Save transcripts", hint: "Keep full text on this device", isOn: $preferences.saveTranscripts)
                        divider
                        ToggleRow("Speak translations aloud", hint: "Play synthesised voice on the other half", isOn: $preferences.autoplay)
                    }
                    group("Accessibility") {
                        ToggleRow("Haptic confirmation", hint: "Buzz on press and release", isOn: $preferences.haptics)
                        divider
                        ToggleRow("Larger transcript text", hint: "Increase live text size by 20%", isOn: $preferences.largeText)
                    }
                    SubscriptionSection(entitlements: entitlements, onUpgrade: onUpgrade)
                    OfflineLanguagesSection(offline: offline, onStart: onStartOffline)
                    group("Legal") {
                        LinkRow("Privacy policy", hint: "How microphone audio and transcripts are handled", url: privacyURL)
                        divider
                        LinkRow("Terms of use", hint: "Apple's standard licensed application terms",
                                url: URL(string: "https://www.apple.com/legal/internet-services/itunes/dev/stdeula/")!)
                    }
                }
                .padding(.horizontal, 20)
                .padding(.top, 6)
                .padding(.bottom, 96)
            }
        }
        .background(VocaTheme.ground.ignoresSafeArea())
    }

    private var divider: some View { Rectangle().fill(VocaTheme.ink.opacity(0.06)).frame(height: 1) }

    private func group<Content: View>(_ title: String, @ViewBuilder content: () -> Content) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            SectionHeader(title)
            VocaCard { VStack(spacing: 0, content: content) }
        }
    }
}

/// JSX `SubscriptionSection`. Cancellation is deliberately not offered: the
/// store requires it in the user's account.
public struct SubscriptionSection: View {
    @ObservedObject var entitlements: EntitlementStore
    let onUpgrade: () -> Void

    public init(entitlements: EntitlementStore, onUpgrade: @escaping () -> Void) {
        self.entitlements = entitlements
        self.onUpgrade = onUpgrade
    }

    public var body: some View {
        let pro = entitlements.isPro
        let left = entitlements.secondsLeft
        VStack(alignment: .leading, spacing: 0) {
            SectionHeader("Subscription")
            VocaCard {
                VStack(spacing: 0) {
                    VStack(alignment: .leading, spacing: 4) {
                        HStack(alignment: .firstTextBaseline, spacing: 8) {
                            Text(pro ? "Voca Pro" : "Free").vocaDisplay(20)
                            if pro {
                                Text("\(VocaFormat.minutes(left)) left")
                                    .vocaMicroLabel(10, color: left > 0 ? VocaTheme.greenDeep : VocaTheme.error, em: 0.1)
                            }
                        }
                        Text(pro
                             ? "\(PurchasesService.planMinutes) minutes of cloud translation each month, then on-device translation stays available."
                             : "On-device translation, free and unlimited, for the language pairs your phone can do without a network.")
                            .vocaBody(12.5, color: VocaTheme.inkMute).lineSpacing(3)
                            .fixedSize(horizontal: false, vertical: true)
                        if pro {
                            ProgressBar(fraction: Double(max(0, left)) / Double(PurchasesService.planMinutes * 60))
                                .padding(.top, 7)
                        }
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.horizontal, 16)
                    .padding(.vertical, 15)
                    Rectangle().fill(VocaTheme.ink.opacity(0.06)).frame(height: 1)
                    Button(action: onUpgrade) {
                        HStack(spacing: 12) {
                            Text(pro ? "Manage or restore purchase" : "Upgrade to Pro — \(entitlements.displayPrice)/mo")
                                .vocaBody(14.5, weight: .semibold, color: pro ? VocaTheme.ink : VocaTheme.violet)
                            Spacer(minLength: 0)
                            Icon(.chevron, size: 16, color: VocaTheme.ink.opacity(0.3))
                        }
                        .padding(.horizontal, 16)
                        .padding(.vertical, 14)
                        .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                }
            }
        }
    }
}

/// JSX `OfflineTranslationSection`: status against English (the pivot), with
/// in-app installs via `prepare` (both directions).
public struct OfflineLanguagesSection: View {
    let offline: OfflineAvailability
    let onStart: () -> Void

    @State private var statuses: [String: PairStatus]?
    @State private var busy: String?
    @State private var error: String?

    public init(offline: OfflineAvailability, onStart: @escaping () -> Void) {
        self.offline = offline
        self.onStart = onStart
    }

    private var supported: [String] { Languages.offlineCodes.filter { $0 != "en" } }

    public var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            SectionHeader("Offline translation")
            VocaCard {
                VStack(spacing: 0) {
                    Text("Download a language to translate it to and from English with no connection. Each is a few hundred MB, so use Wi-Fi. Cantonese has no offline model on any phone and always needs a connection.")
                        .vocaBody(12.5, color: VocaTheme.ink.opacity(0.62)).lineSpacing(3)
                        .fixedSize(horizontal: false, vertical: true)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .padding(.horizontal, 15).padding(.vertical, 13)
                    ForEach(supported, id: \.self) { code in
                        Rectangle().fill(VocaTheme.ink.opacity(0.06)).frame(height: 1)
                        row(code)
                    }
                }
            }
            Button(action: onStart) {
                HStack(spacing: 8) {
                    Icon(.mic, size: 17, color: VocaTheme.ink)
                    Text("Start an offline session").vocaBody(15, weight: .semibold)
                }
                .frame(maxWidth: .infinity).frame(height: 52)
                .background(VocaTheme.surface)
                .clipShape(RoundedRectangle(cornerRadius: VocaTheme.cardRadius, style: .continuous))
                .overlay(RoundedRectangle(cornerRadius: VocaTheme.cardRadius, style: .continuous).stroke(VocaTheme.ink.opacity(0.14)))
            }
            .buttonStyle(.plain)
            .padding(.top, 10)
            if let error {
                Text(error).vocaBody(12, color: VocaTheme.error).padding(.top, 6)
            }
        }
        .task { await refresh() }
    }

    private func row(_ code: String) -> some View {
        let lang = Languages.get(code)
        let status = statuses?[code]
        return HStack(spacing: 12) {
            Text(lang.flag).font(.system(size: 17))
            Text(lang.label).vocaBody(14.5, weight: .semibold)
            Spacer(minLength: 0)
            if statuses == nil {
                Text("Checking…").vocaBody(12, color: VocaTheme.ink.opacity(0.4))
            } else {
                switch status {
                case .installed:
                    Text("On device").vocaBody(12, weight: .semibold, color: VocaTheme.greenDeep)
                case .supported:
                    Button { install(code) } label: {
                        Text(busy == code ? "Downloading…" : "Download")
                            .vocaBody(12.5, weight: .semibold, color: busy == code ? VocaTheme.ink.opacity(0.4) : VocaTheme.violet)
                    }
                    .buttonStyle(.plain)
                    .disabled(busy != nil)
                default:
                    Text("Not available").vocaBody(12, color: VocaTheme.ink.opacity(0.35))
                }
            }
        }
        .padding(.horizontal, 15)
        .padding(.vertical, 12)
    }

    private func refresh() async {
        var result: [String: PairStatus] = [:]
        for code in supported {
            result[code] = await offline.pairStatus("en", code)
        }
        statuses = result
    }

    private func install(_ code: String) {
        busy = code
        error = nil
        Task {
            do {
                _ = try await offline.prepare("en", code)
            } catch {
                self.error = "Could not download \(Languages.get(code).label). Check your connection and try again."
            }
            await refresh()
            busy = nil
        }
    }
}

File: native/android/app/src/main/java/com/vocare/translate/app/ui/SettingsScreen.kt
package com.vocare.translate.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Switch
import androidx.compose.material3.SwitchDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.vocare.translate.app.store.SettingKey
import com.vocare.translate.app.store.Tier
import com.vocare.translate.core.model.Language
import com.vocare.translate.core.offline.OfflineLanguages
import com.vocare.translate.core.theme.VocaTheme

private typealias SC = VocaTheme.Colors

/** Legal links live in one place so Settings and the store listing cannot drift. */
object LegalLinks {
    const val PRIVACY = "https://vocare.cx/privacy"
    const val TERMS = "https://vocare.cx/terms"
    const val SUPPORT = "https://vocare.cx/support"
}

/**
 * Settings (CONTRACT §5): session toggles, the subscription section, offline
 * language packs with in-app install buttons, and the legal links.
 */
@Composable
fun SettingsScreen(
    state: AppUiState,
    onToggle: (SettingKey, Boolean) -> Unit,
    onManageSubscription: () -> Unit,
    onRestore: () -> Unit,
    onInstall: (String) -> Unit,
    onOpenLink: (String) -> Unit,
    onBack: () -> Unit,
    banner: @Composable () -> Unit = {},
) {
    Column(
        Modifier.fillMaxSize().background(SC.Ground).verticalScroll(rememberScrollState()),
    ) {
        VocaTopBar("Settings", onBack = onBack)
        Column(Modifier.padding(horizontal = 20.dp)) {
            banner()

            SectionLabel("Session")
            VocaCard {
                ToggleRow("Show live notes", "Keep partial transcripts on screen while someone speaks", state.settings.notes) {
                    onToggle(SettingKey.NOTES, it)
                }
                ToggleRow("Save transcripts", "Keep finished sessions in History on this device", state.settings.saveTranscripts) {
                    onToggle(SettingKey.SAVE, it)
                }
                ToggleRow("Speak translations", "Play the translation aloud when it is ready", state.settings.speakAloud) {
                    onToggle(SettingKey.AUTOPLAY, it)
                }
                ToggleRow("Haptics", "Vibrate when a turn starts and ends", state.settings.haptics) {
                    onToggle(SettingKey.HAPTICS, it)
                }
                ToggleRow("Large text", "Bigger transcript type on the live screen", state.settings.largeText, last = true) {
                    onToggle(SettingKey.LARGE, it)
                }
            }

            SectionLabel("Subscription")
            VocaCard {
                val pro = state.snapshot.tier == Tier.PRO
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Column(Modifier.weight(1f)) {
                        Text(if (pro) "Voca Pro" else "Free", style = VocaTheme.Type.BodySemi, color = SC.Ink)
                        Spacer(Modifier.height(3.dp))
                        Text(
                            if (pro) {
                                "${state.snapshot.secondsLeft / 60} of ${state.snapshot.balance?.secondsTotal?.div(60) ?: 0} " +
                                    "cloud minutes left this period"
                            } else {
                                "On-device translation, unlimited"
                            },
                            style = VocaTheme.Type.Body.copy(fontSize = 12.5.sp),
                            color = SC.TextMuted,
                        )
                    }
                    TierChip(if (pro) "Pro" else "Free", pro = pro, onClick = onManageSubscription)
                }
                Spacer(Modifier.height(12.dp))
                VocaSecondaryButton(if (pro) "Manage subscription" else "See Voca Pro", onManageSubscription)
                Spacer(Modifier.height(8.dp))
                VocaSecondaryButton("Restore purchases", onRestore, enabled = state.storeAvailable && !state.purchaseBusy)
                state.purchaseMessage?.let {
                    Spacer(Modifier.height(8.dp))
                    Text(it, style = VocaTheme.Type.Body, color = SC.ADeep)
                }
            }

            SectionLabel("On-device languages")
            VocaCard {
                Text(
                    "Downloaded language packs translate with no network. Each is around 30 MB.",
                    style = VocaTheme.Type.Body.copy(fontSize = 12.5.sp),
                    color = SC.TextMuted,
                )
                Spacer(Modifier.height(10.dp))
                OfflineLanguages.CODES.forEach { code ->
                    OfflineRow(
                        language = Language.of(code),
                        installed = code in state.offlineInstalled,
                        installing = state.installing == code,
                        onInstall = { onInstall(code) },
                    )
                }
            }

            SectionLabel("About")
            VocaCard {
                LinkRow("Privacy policy") { onOpenLink(LegalLinks.PRIVACY) }
                LinkRow("Terms of service") { onOpenLink(LegalLinks.TERMS) }
                LinkRow("Support", last = true) { onOpenLink(LegalLinks.SUPPORT) }
            }
            Spacer(Modifier.height(28.dp))
        }
    }
}

@Composable
private fun SectionLabel(text: String) {
    Spacer(Modifier.height(18.dp))
    MicroLabel(text, color = SC.InkMute)
    Spacer(Modifier.height(8.dp))
}

@Composable
private fun ToggleRow(
    title: String,
    subtitle: String,
    checked: Boolean,
    last: Boolean = false,
    onChange: (Boolean) -> Unit,
) {
    Row(Modifier.fillMaxWidth().padding(vertical = 8.dp), verticalAlignment = Alignment.CenterVertically) {
        Column(Modifier.weight(1f)) {
            Text(title, style = VocaTheme.Type.BodySemi, color = SC.Ink)
            Spacer(Modifier.height(2.dp))
            Text(subtitle, style = VocaTheme.Type.Body.copy(fontSize = 12.sp), color = SC.TextMuted)
        }
        Switch(
            checked = checked,
            onCheckedChange = onChange,
            colors = SwitchDefaults.colors(
                checkedTrackColor = SC.A,
                checkedThumbColor = Color.White,
                uncheckedTrackColor = SC.SwitchOff,
                uncheckedThumbColor = Color.White,
            ),
        )
    }
    if (!last) Box(Modifier.fillMaxWidth().height(1.dp).background(SC.Divider))
}

@Composable
private fun OfflineRow(
    language: Language,
    installed: Boolean,
    installing: Boolean,
    onInstall: () -> Unit,
) {
    Row(Modifier.fillMaxWidth().padding(vertical = 7.dp), verticalAlignment = Alignment.CenterVertically) {
        Text(language.flag, fontSize = 18.sp)
        Spacer(Modifier.width(10.dp))
        Text(language.label, style = VocaTheme.Type.Body, color = SC.Ink, modifier = Modifier.weight(1f))
        when {
            installed -> MicroLabel("Installed", color = SC.BDeep)
            installing -> MicroLabel("Downloading…", color = SC.ADeep)
            else -> Box(
                modifier = Modifier
                    .clip(VocaTheme.Radii.ButtonShape)
                    .background(SC.Surface2)
                    .clickable(onClick = onInstall)
                    .padding(horizontal = 14.dp, vertical = 7.dp),
            ) { MicroLabel("Install", color = SC.ADeep) }
        }
    }
}

@Composable
private fun LinkRow(title: String, last: Boolean = false, onClick: () -> Unit) {
    Row(
        Modifier.fillMaxWidth().clickable(onClick = onClick).padding(vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(title, style = VocaTheme.Type.Body, color = SC.Ink, modifier = Modifier.weight(1f))
        Text("›", style = VocaTheme.Type.BodySemi.copy(fontSize = 18.sp), color = SC.TextFaint)
    }
    if (!last) Box(Modifier.fillMaxWidth().height(1.dp).background(SC.Divider))
}
